docs: ICM lifecycle and API compatibility governance draft (#497)
camaraproject/ReleaseManagement: documentation/SupportingDocuments/icm-lifecycle-and-compatibility.md

# ICM Lifecycle and API Compatibility Governance

**Version:** Draft 1 (2026-04-21)
**Status:** Draft for Release Management WG discussion.
**Scope:** Response to the ICM WG request to Release Management (per [ICM#324](https://github.com/camaraproject/IdentityAndConsentManagement/issues/324), [ICM#340](https://github.com/camaraproject/IdentityAndConsentManagement/issues/340), [ReleaseManagement#351](https://github.com/camaraproject/ReleaseManagement/issues/351)) to define governance for ICM version evolution and its dependencies with CAMARA API versions.

---

## 1. Scope and Purpose

> This guideline defines lifecycle management of Identity and Consent Management (ICM) versions and how CAMARA APIs declare and maintain compatibility with them. It answers the question posed by the ICM Working Group to Release Management: under what rules can API versions and ICM versions co-evolve while still providing clear compatibility guarantees to API Consumers?

Reviewer: ```suggestion
This guideline defines lifecycle management of Identity and Consent Management (ICM) versions and how CAMARA APIs declare and maintain compatibility with them. It defines under what rules API versions and ICM versions can co-evolve while still providing clear compatibility guarantees.
```

Author: Applied in subsequent commits. §1 Scope and Purpose carries equivalent wording.

The guideline recognizes operational reality: API Providers offer the same API version against multiple ICM versions, and API Consumers require an API version and an ICM version as part of a single usage contract — if either aspect changes in a way that affects them, their implementation must adapt.

## 2. Glossary

Terms defined in the CAMARA Commonalities glossary (API, API Consumer, API Provider, meta-release, semantic versioning, scope, etc.) are not repeated here. This section defines terms specific to this guideline.

> - **ICM-compatibility**: the guarantee that an API version's specification is valid against a given ICM version — its scope format, `securitySchemes` syntax, schemas, and operations use constructs defined by that ICM version. ICM-compatibility is a property of the API specification relative to an ICM version.

Reviewer: - Change "API specification" to "API definition" everywhere.
- Change "API version's specification" to API version's definition"

Author: Applied in subsequent commits. "API specification" → "API definition" applied throughout subsequent commits.

> - **ICM major version**: the major component of an ICM semantic version (e.g., "1" in "1.y.z"). Starting with ICM 1.0.0, major-version increments indicate breaking changes for API Consumers or API specifications that cannot be expressed additively.

Reviewer: ```suggestion
- **ICM version**: A Semantic Versioning (SemVer 2.0) compliant version number with major, minor and patch components, together identifying a specific set of ICM artefacts. Starting with ICM 1.0.0, ICM major version increments indicate breaking changes for API definitions or API deployments.
```

Author: Applied in subsequent commits. Glossary §2 entry for "ICM version" carries equivalent wording.

> - **Supported / Deprecated / Retired / Revoked**: the four lifecycle states for an ICM version (see §6). States apply per ICM version, with the ICM major version as the primary grain.

Reviewer: ```suggestion
- **ICM lifecycle states**: Supported / Deprecated / Retired / Revoked are the four possible lifecycle states of an ICM version (see §6). Lifecycle states apply per ICM version.
```

Author: Applied in subsequent commits. Glossary §2 entry for "ICM lifecycle states" carries equivalent wording.

> - **Preferred**: a designation — not a state — applied to exactly one Supported ICM major version at any given time, identifying it as the current baseline for newly published CAMARA API versions (see §6).

Reviewer: I don't think this needs to apply only to major versions. Adding a minor feature is an ICM minor update but it would become the preferred one. 
So suggest to drop the word "major".

Author: Do we need the concept of "Preferred" at all? Otherwise we need to differentiate "preferred" for design time (the one defined as the target ICM / Commonalities versions for a meta-release, and the one deployed at API providers. The latter one is a different one at the time of the meta-release, and especially at the time between Signal and Sync meta-releases. I would rather see a kind of recommended deployment timeline for API providers: between Signal and Sync prepare for the new ICM version, after Sync meta-release offer the version within X months in production. WDYT? 

One concept less would simplify the document.

Reviewer: Let's drop the Preferred label. 

On the recommended deployment (or rather upgrade ?) time for API deployments: I think this should be decoupled from meta-releases as such. It is based on lifecycle state and would correspond to the time period to transition a Deprecated  ICM version to Retired state. 
However, the decision on ICM version retirement will probably be done as part of each Signal meta-release process, but is not necessarily linked to the time-line between meta-releases.
The latter is only valid for API definitions, not for API deployments.

Author: done in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/d934e6aadbeb26435309e5b9a0ae822637864196

> - **Preferred**: a designation — not a state — applied to exactly one Supported ICM major version at any given time, identifying it as the current baseline for newly published CAMARA API versions (see §6).

Reviewer: ```suggestion
- **Preferred**: a label — not a state — applied to exactly one Supported ICM version at any given time, identifying it as the current baseline for new API versions starting development from scratch (v0.1.0) (see §6).
```

Author: "Exactly one" would not fit to different to the differentiation in design and deployment compatibility. And for a meta-release there is no "preferred", but a mandatory ICM version (at least if was in the past).

Reviewer: True that preferred only applied to design ICM compatibility. Another reason to drop it.

Author: done in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/d934e6aadbeb26435309e5b9a0ae822637864196

- **Exception (waiver)**: a time-bound, governance-approved authorization that permits a specific (API version, ICM version) pair outside the normal ICM-compatibility guarantee.
- **Compatibility matrix**: the derived artifact listing which (API version, ICM version) pairs are ICM-compatible at a given point in time (see §12).

> 

Reviewer: ```suggestion
- **ICM governance**: the decision process to transition an ICM version to a different lifecycle state (see §6). 

```

## 3. ICM-compatibility

> ICM-compatibility, as defined in the glossary, applies to an API version's specification. It is established by the API Sub Project at API publication time via the `x-camara-min-icm-version` field (§7) and verified by governance through the compatibility matrix (§12).

Reviewer: suggest to rename to `x-camara-min-icm` or `x-camara-minimum-icm` to be consistent with `x-camara-commonalities` which does not have "version" in its name. It also makes it easier about  `x-camara-min-icm` version (instead of " `x-camara-min-icm-version` version").

The value of both `x-camara- ... ` fields should be a valid CAMARA SemVer 2.0 version.

Author: agree on dropping "version". just not happy with with the abbreviation "icm" which is not known to "non CAMARA experts". We need a kind of description or descriptive comment for this field then.

Author: done in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/d934e6aadbeb26435309e5b9a0ae822637864196

> ICM-compatibility, as defined in the glossary, applies to an API version's specification. It is established by the API Sub Project at API publication time via the `x-camara-min-icm-version` field (§7) and verified by governance through the compatibility matrix (§12).

Reviewer: ```suggestion
ICM-compatibility, as defined in the glossary, applies to an API version's OAS definition. It is established by the API Sub Project at API public release time via the `x-camara-min-icm-version` field (§7) and maintained in the compatibility matrix (§12).
```

Author: Applied in subsequent commits. References outdated field name `x-camara-min-icm-version`. The -version suffix was dropped in the V2 working merge per your later L19 / L20 comments; current field is `x-camara-min-icm`.

> API Consumers use an API in combination with an API Provider's ICM implementation. To successfully integrate, a Consumer must implement the authentication and authorization behaviors — flows, grants, assertions, token processing — defined by the ICM version the Provider implements. This is independent of the API version being consumed: an ICM change that tightens authentication rules affects Consumers interacting with any Provider who moves to that ICM version, regardless of which APIs they consume. ICM-compatibility of the API specification and the Consumer's implementation readiness are separate concerns — both must be satisfied for a Consumer to successfully integrate.

Reviewer: ```suggestion
API Consumers use an API in combination with an API Provider's ICM implementation. To successfully integrate, a Consumer must implement the authentication and authorization behaviors — flows, grants, assertions, token processing — defined by the ICM version the Provider implements. These ICM version aspects are **independent** of the API version being consumed: an ICM change that tightens authentication rules affects Consumers interacting with any Provider who moves to that ICM version, regardless of which APIs they consume. ICM-compatibility of the API specification and the Consumer's implementation readiness are separate concerns — both must be satisfied for a Consumer to successfully integrated.
```

Reviewer: in this doc, we define ICM-compatibility applicable only to an API version (definitions).
Need to add, as discussed, ICM-compatibility for API deployments as well.

Author: Applied in subsequent commits. API Consumer responsibilities covered in current §7.2 (post-renumber); independence of ICM and API versions covered in §3.3.

An ICM change may affect the API specification (triggering an API release to restore ICM-compatibility), the API Consumer implementation (requiring Consumer-side updates), or both. The semver discipline in §5 applies to both kinds of breaking changes.

> ## 4. Core rule

Reviewer: ```suggestion
## 4. Minimum ICM version
```

Author: No longer applicable. §4 was merged into §6.1 in commit 0ef3907 ("Minimum ICM version" is now §6.1 — see r3260364170 thread).

> An API version declares its minimum ICM version via `x-camara-min-icm-version`. This floor expresses the oldest ICM version whose constructs the API specification relies on and which is Supported at the release time of the API. Subject to ICM's semver discipline (§5), the API specification is guaranteed ICM-compatible with any ICM version at or above that floor, at least within the same ICM major.

Reviewer: ```suggestion
An API version declares its minimum ICM version via `x-camara-min-icm-version` field in its API definition. This floor expresses the oldest ICM version that the API definition relies on and which is Supported at the time of release of the API version. Subject to ICM's SemVer discipline (§5), the API version is guaranteed ICM-compatible with any ICM version at or above that floor, at least within the same major ICM version.
```

Author: Superseded by your later input. Floor declaration now lives in §6.1 after the §4 merge; current wording covers this. Outdated `-version` field name.

When ICM crosses a major-version boundary, each API specification is assessed by governance at that boundary. API specifications that do not rely on ICM-version-specific constructs may be declared ICM-compatible with the new major through an explicit governance decision recorded in the compatibility matrix. Specifications that do rely on such constructs must be re-released as a new API major targeting the new ICM major. §7.3 describes an optional field that allows an API specification to declare compatibility with more than one ICM major explicitly.

> Concretely: `x-camara-min-icm-version: 1.2.0` declared in an API specification means the API depends on ICM 1.x features from version 1.2.0 onward. The specification is guaranteed ICM-compatible with any Supported or Deprecated ICM version ≥ 1.2.0 in the 1.x major. Whether it remains ICM-compatible with a future 2.x version is a governance assessment made at the time of the ICM 2.0 release.

Reviewer: ```suggestion
Example: `x-camara-min-icm-version: 1.2.0` declared in an API specification means the API depends on ICM 1.x features from version 1.2.0 onward. The specification is guaranteed ICM-compatible with any Supported or Deprecated ICM version ≥ 1.2.0 in the 1.x major. Whether it remains ICM-compatible with a future 2.x version is a governance assessment made at the time of the ICM 2.0 release.
```

Author: Applied in subsequent commits. Example now in §6.1; current wording uses `x-camara-min-icm` (no -version) per the later naming decision.

> ## 5. Prerequisite: ICM semver commitment at 1.0.0

Reviewer: ```suggestion
## 5. ICM SemVer commitment (starting at 1.0.0)
```

Author: No longer applicable. §5 was renamed to "ICM versioning"; current §4 has the same heading.

> This guideline assumes ICM adopts strict semver starting with its 1.0.0 release. The discipline applies to two distinct classes of breaking changes:

Reviewer: ```suggestion
This guideline assumes ICM adopts strict SemVer versioning starting with its 1.0.0 release. The discipline applies to two distinct classes of information changes:
```

Author: Applied in subsequent commits. §4 (current numbering) "ICM versioning" carries the two-classes framing.

- **Major versions (1.x → 2.0)**: reserved for breaking changes that cannot be expressed additively. These may affect ICM-compatibility (for example, replacement of a mandatory security schema, an incompatible scope format, removal of a required claim), affect API Consumer implementations (for example, mandatory new authentication rules, an assertion lifetime cap that rejects existing Consumers), or both. Majors should be deliberately rare and driven by security or regulatory necessity.
- **Patch versions (1.2.3 → 1.2.4)**: editorial or defect corrections with no impact on either ICM-compatibility or API Consumer implementations.

An ICM change that invalidates existing Consumer implementations is a major-version event even if no API specification requires re-release. Semver discipline is not satisfied by preserving ICM-compatibility alone.

The transition to 1.0.0 should coincide with a scope-baseline review of the ICM documents, declaring the then-current definitions as the stable starting point. Pre-1.0 versions are handled by the Legacy section (§13).

> ## 6. ICM lifecycle governance

Reviewer: ```suggestion
## 6. ICM lifecycle states and governance
```

Author: Applied in subsequent commits. Current §5 heading is "ICM version lifecycle states and governance".

### 6.1 State model

> Each ICM version moves through lifecycle states. States apply per ICM version, with the ICM major version as the primary grain: a new major enters the Supported state at its public release, and sub-major versions (minor and patch) inherit their major's state by default. Governance MAY explicitly assign a different state to a specific sub-major version.

Reviewer: Change "ICM major version" to "major ICM version"

Author: Applied in subsequent commits. "major ICM version" used consistently throughout.

> | State | Meaning |

Reviewer: ```suggestion
| State | ICM version | API version |  API deployment |
```

Author: Applied in subsequent commits. §5.1 table has columns "State | ICM version | API version | API deployment".

> |---|---|

Reviewer: ```suggestion
|---|---|---|---|
```

Author: Applied in subsequent commits. Table separator in §5.1 has 4 pipes matching the header.

> | **Supported** | Active; APIs may declare and operate against this version |

Reviewer: ```suggestion
| **Supported** | is OK to use | may be used as `x-camara-min-icm` | may operate against this ICM version |
```

Author: Applied in subsequent commits. Supported row carries the suggested wording in §5.1.

> | **Deprecated** | Sunset announced; migration window active; no new APIs should declare this version |

Reviewer: ```suggestion
| **Deprecated** | Sunset announced: ICM migration window active |  do not use as `x-camara-min-icm` | may operate against this ICM version |
```

Author: Applied in subsequent commits. Deprecated row carries equivalent wording in §5.1.

> | **Retired** | Terminal (planned end of life); APIs bound to a Retired version are no longer CAMARA-compliant and must transition or be Retired at the API level |

Reviewer: ```suggestion
| **Retired** | Terminal (planned end of life) | is no longer ICM-compatible and must either evolve to a Supported ICM version or be transitioned to Retired | must migrate to an ICM-compatible deployment with a a Supported ICM version |
```

Author: Applied in subsequent commits. Retired row carries equivalent wording in §5.1.

> | **Revoked** | Terminal (exceptional); the version has been replaced by a later patch or minor due to defects or security issues and MUST NOT be used. Governance names the replacement version. |

Reviewer: ```suggestion
| **Revoked** | Terminal (exceptional): the version has been replaced by a later minor or patch version due to defects or security issues and MUST NOT be used. Governance names the replacement version. | MUST re-establish ICM-compatibility with a Supported ICM version | MUST migrate to Supported ICM version and ICM-compatible deployment |
```

Author: Applied in subsequent commits. Revoked row carries equivalent wording in §5.1; broadened to "another (earlier or later)" per your L110 in commit 5bca022.

> The planned state-transition sequence is Supported → Deprecated → Retired. Revoked is entered as an exceptional action outside this sequence and requires an explicit replacement version.

Reviewer: Should we make "Revoked" a label, rather than a state ? I think it could simplify a bit.

Author: Why? It is clear lifecycle state ... the difference to "Retired" is that is has a clear replacement version (typical case: a public minor release introduce accidentally a breaking change - corrected in a patch release which replaces the broken, revoked release).

Reviewer: Having the replacement version was also the case for the "Preferred" label, so there was some analogy there.  

All the resulting behavior (for API definitions and for deployments) for Revoked is the same as for Retired, it is just due to another decision, so the documentation of state transition management (nd the understanding of it) will be a bit more cumbersome. 
But OK if you want to keep it.

Author: Kept as a state per the 2026-05-13 discussion — the lifecycle implications for API definitions and API deployments are part of the formal state model, not a side-label.

Author: Kept as a state per the 2026-05-13 discussion — the lifecycle implications for API definitions and API deployments are part of the formal state model, not a side-label.

> **Preferred** is not a lifecycle state but a designation applied to exactly one Supported ICM major version at any given time, identifying it as the current baseline for newly published CAMARA API versions. The Preferred designation is assigned by governance and MAY be reassigned to a newer Supported major at its public release or at a later governance decision. Reassignment of the Preferred designation is an explicit governance action, not automatic. A major that loses the Preferred designation remains in the Supported state until governance moves it to Deprecated.

Reviewer: ```suggestion
**Preferred** is not a lifecycle state but a label applied to exactly one Supported ICM version at any given time, identifying it as the current baseline for newly published API versions. The Preferred label is assigned by governance and MAY be reassigned to a newer Supported ICM version either at its public release or at a later governance decision. Reassignment of the Preferred label is not automatic. An ICM version that loses the Preferred label remains in the Supported state until governance transitions it to Deprecated.
```

Author: No longer applicable. Preferred concept dropped entirely in commit d934e6a per your later input.

> Governance MAY mark specific minor or patch versions within a Supported major as Deprecated, Retired, or Revoked when they should no longer be used as a floor — for example, when known ambiguities or defects are resolved in later minors, or when a critical defect requires replacing a specific patch version. Such per-version markings do not introduce incompatibility for API clients — later minors remain compatible with earlier ones by semver — but they influence the "lowest Supported minor" used in §7.2 and are relevant for API Providers.

Reviewer: ```suggestion
For Supported ICM versions, governance MAY transition specific minor or patch ICM versions with the same major version number as Deprecated, Retired, or Revoked when they should no longer be used as a floor — for example, when known ambiguities or defects are resolved in a later minor ICM version, or when a critical defect requires replacing a specific patch ICM version. Such per-version transitions do not impact ICM-compatibility of API versions or API deployments — later minor ICM versions remain compatible with earlier ones by SemVer — but they influence the "lowest Supported minor" used in §7.2 and are relevant for API deployments.
```

Reviewer: this paragraph needs some rewrites with the new ICM-compatibility definition. Not sure about the "lowest Supported minor" (need to check the paragraph).

Author: Applied in subsequent commits. Per-version transitions covered in current §5.2 (post-restructure).

# ICM Lifecycle and API Compatibility Governance

**Version:** Draft 1 (2026-04-21)
**Status:** Draft for Release Management WG discussion.
**Scope:** Response to the ICM WG request to Release Management (per [ICM#324](https://github.com/camaraproject/IdentityAndConsentManagement/issues/324), [ICM#340](https://github.com/camaraproject/IdentityAndConsentManagement/issues/340), [ReleaseManagement#351](https://github.com/camaraproject/ReleaseManagement/issues/351)) to define governance for ICM version evolution and its dependencies with CAMARA API versions.

---

## 1. Scope and Purpose

> This guideline defines how Identity and Consent Management (ICM) evolves as an independent Working Group and how CAMARA APIs declare and maintain compatibility with ICM versions. It answers the question posed by the ICM Working Group to Release Management: under what rules can API versions and ICM versions co-evolve while still providing clear compatibility guarantees to API Consumers, aggregators, and certification authorities?

Reviewer: ```suggestion
This guideline defines lifecycle management of Identity and Consent Management (ICM) versions and how CAMARA APIs declare and maintain compatibility with ICM version. It answers the question posed by the ICM Working Group to Release Management: under what rules can API versions and ICM versions co-evolve while still providing clear compatibility guarantees to API Consumers ?
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/eff45e532aa174f4b3941c8131fbb6a0b88c2dc9

> The guideline recognizes operational reality: providers offer the same API version against multiple ICM versions, aggregators depend on this flexibility, and API Consumers combine an API version and an ICM version into a single usage contract — if either side changes in a way that affects them, their integration must adapt.

Reviewer: ```suggestion
The guideline recognizes operational reality: providers offer the same API version against multiple ICM versions, aggregators depend on this flexibility, and API Consumers require an API version and an ICM version to be part of a single usage contract — if either aspect changes in a way that affects them, their implementation must adapt.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/eff45e532aa174f4b3941c8131fbb6a0b88c2dc9

## 2. Glossary

Terms defined in the CAMARA Commonalities glossary (API, API Consumer, API Provider, meta-release, semantic versioning, scope, etc.) are not repeated here. This section defines terms specific to this guideline.

> - **ICM↔API contract**: the set of ICM-defined rules that shape how APIs are specified — scope format, `securitySchemes` syntax, sub claim format, token structure assumed by the API. Distinct from the contract between API Consumer and API (operations, schemas) and the contract between API Consumer and Authorization Server (flows, grants, assertions).

Reviewer: ```suggestion
- **ICM-compatibility**: the guarantee that an API version respects the set of rules defined by a given ICM version — auth flows, grants, assertions, scope format, `securitySchemes` syntax, sub claim format, token structure assumed by the API, relevant operations and schemas.
```

Author: Partially applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/eff45e532aa174f4b3941c8131fbb6a0b88c2dc9

In this draft ICM-compatibility is scoped to the API specification (scope format, securitySchemes syntax, schemas, operations). Authentication flows, grants, assertion format, and token processing live on the API Consumer's implementation side — they are governed by the same semver discipline (§5) but are a separate concern, not part of ICM-compatibility itself. This is explained in §3 paragraph 2.

Reviewer: Forgot to answer the comments here. 

I guess here is where our opinions differ: For me, ICM-compatibility also should cover the supported auth flows etc. definitions described in the ICM's version documentation (impacting client and server side), so it scopes beyond just the API definition only.

> - **Orthogonal API**: an API whose specification is not affected by differences between ICM major versions — its `securitySchemes`, scopes, and operational semantics are valid under any ICM major currently in the Supported tier. The majority of current CAMARA APIs are orthogonal.

Reviewer: not sure we need to introduce this concept nor to have a definition for this. Just use ICM-compatibility across major ICM versions.

I would prefer to drop it

Author: Dropped in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/eff45e532aa174f4b3941c8131fbb6a0b88c2dc9

> - **ICM major version**: the major component of an ICM semantic version (e.g., "1.x"). Starting with ICM 1.0.0, major-version increments indicate client-facing changes that cannot be expressed additively.

Reviewer: ```suggestion
- **ICM major version**: the major component of an ICM semantic version (e.g., "1" on "1.y.z"). Starting with ICM 1.0.0, major-version increments indicate client-facing changes that cannot be expressed additively (e.g. breaking changes for API consumers)
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/eff45e532aa174f4b3941c8131fbb6a0b88c2dc9

> - **Supported / Deprecated / Retired**: the three lifecycle tiers for an ICM major version (see §6).

Reviewer: ```suggestion
- **Supported / Deprecated / Retired**: the three possible lifecycle states for an ICM version (see §6).
```

Author: Partially applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/eff45e532aa174f4b3941c8131fbb6a0b88c2dc9

Four states in the updated draft: Supported / Deprecated / Retired / Revoked. Revoked added as an exceptional terminal state for versions replaced by a later patch or minor due to defects or security issues (with an explicit replacement version named). See §6.1.

> - **Preferred**: a designation — not a tier — applied to exactly one Supported ICM major at a time, identifying it as the current baseline for new API development (see §6).

Reviewer: ```suggestion
- **Preferred**: a designation — not a state — applied to exactly one Supported ICM version at any given time, identifying it as the current baseline for new API development (see §6).
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/eff45e532aa174f4b3941c8131fbb6a0b88c2dc9

> - **Overlap window**: the period during which a previous ICM major remains in the Supported tier after the Preferred designation has moved to a newer major.

Reviewer: this should be rework to just use "Supported state availability timeframe" or something instead of introducing yet another concept.

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/eff45e532aa174f4b3941c8131fbb6a0b88c2dc9

> - **Exception (waiver)**: a time-bound, governance-approved compatibility authorization that permits a specific (API version, ICM version) pair outside the normal rules.

Reviewer: ```suggestion
- **Exception (waiver)**: a time-bound, governance-approved compatibility authorization that permits a specific (API version, ICM version) pair outside the normal ICM-compatibility guarantee.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/eff45e532aa174f4b3941c8131fbb6a0b88c2dc9

- **Compatibility matrix**: the derived artifact listing which (API version, ICM version) pairs are compatible at a given point in time (see §12).

> ## 3. The three contracts

Reviewer: I would like to drop this section and the 3 contracts as not needed IMHO, only the API definition is a contract for API consumers (with x-camara-minimum-icm included).

The API - ICM "contract" or rather ICM-compatibility guarantee should be by API providers who should ensure ICM compatibility of their API implementations

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/eff45e532aa174f4b3941c8131fbb6a0b88c2dc9

| **ICM ↔ API** | AS behavior and API specification | Scope format, `securitySchemes` syntax, sub claim format, token shape the API may rely on | ICM defines; APIs follow |

These three contracts are largely independent:

- An API spec change that does not alter its scope format or `securitySchemes` declarations does not affect contracts 2 or 3.
- An ICM change to flow mechanics (e.g., tightening assertion lifetime) affects contract 2 only — the API is unaffected.
- An ICM change to scope format or `securitySchemes` syntax affects contract 3; API specs written against the prior syntax remain valid against AS implementations that honor that prior syntax, but cannot be published as new specs once the syntax has changed.

**From the API Consumer's perspective, the combined contract is what matters**: they integrate against both a specific API version and a specific ICM version. A change in either side that affects them requires their integration to adapt. The independence of the three contracts is analytical — useful for governance — but does not relieve the API Consumer of having to manage the pair.

## 4. Core rule

> **An API version is bound to the ICM major version whose ICM↔API contract it was written against.**

Reviewer: ```suggestion
**An API version is ICM-compatible with any Supported ICM version declared in its published definition and higher available ICM versions.**

```

Author: Partially applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/eff45e532aa174f4b3941c8131fbb6a0b88c2dc9

The current §4 narrows the guarantee to "at least within the same ICM major" — a breaking change at an ICM major boundary (new scope format, replacement of a mandatory security schema) can invalidate an API specification written against the previous major, so unbounded forward compatibility cannot be assumed. Cross-major compatibility is either assessed by governance (recorded in the matrix) or declared explicitly by the API via §7.3's optional x-camara-icm-compatibility field.

> For APIs whose contract is not affected by differences between ICM majors — *orthogonal APIs*, which make up the large majority of current CAMARA APIs — the binding has no operational effect: such an API remains compatible with any ICM major currently in the Supported tier.

Reviewer: ```suggestion
For APIs who are not impacted by differences between ICM majors —  which make up the large majority of current CAMARA APIs — the binding has no operational effect: such an API remains compatible with any ICM version currently in the Supported state.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/eff45e532aa174f4b3941c8131fbb6a0b88c2dc9 (concept dropped completely)

> For APIs whose ICM↔API contract IS affected (e.g., an API that uses a specific scope format introduced in an ICM major, or relies on an operator-token mechanism), the binding is explicit and operational: the API can only be served by AS implementations that honor its ICM major's contract.

Reviewer: ```suggestion
For APIs whose ICM-compatibility is impacted (e.g., an API that uses a specific scope format introduced in a new ICM version, or relies on an operator-token mechanism), its compatibility is explicitly expressed in the API version's x-camara-minimum-icm field and operational: the API version can only be served by ICM implementations that provide this ICM version.
```

Author: Partially applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/eff45e532aa174f4b3941c8131fbb6a0b88c2dc9

§4 expresses this more permissively: the guarantee holds within the same ICM major and is extended to newer majors only by governance assessment or by the optional §7.3 declaration. "Only this ICM version" is the narrow case; within-major forward compatibility is the default.

> Concretely: `x-camara-min-icm-version: 1.2.0` declared in an API spec means the API was written against the ICM 1.x ICM↔API contract and requires features introduced from ICM version 1.2.0 onward within that major.

Reviewer: ```suggestion
Concretely: `x-camara-min-icm-version: 1.2.0` declared in an API version means this API version was written against the features introduced from ICM version 1.2.0 and onward.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/eff45e532aa174f4b3941c8131fbb6a0b88c2dc9

---

## 1. Scope and Purpose

This guideline defines lifecycle management of Identity and Consent Management (ICM) versions and how CAMARA APIs declare and maintain compatibility with them. It defines under what rules API versions and ICM versions can co-evolve while still providing clear compatibility guarantees.

The guideline recognizes operational reality: API Providers offer the same API version against multiple ICM versions, and API Consumers require an API version and an ICM version as part of a single usage contract — if either aspect changes in a way that affects them, their implementation must adapt.

## 2. Glossary

Terms defined in the CAMARA Commonalities glossary (API, API Consumer, API Provider, meta-release, semantic versioning, scope, etc.) are not repeated here. This section defines terms specific to this guideline.

> - **ICM design info**: information defined by an ICM version that applies to API definitions at API design time — scope format, `securitySchemes` syntax, mandatory `info.description` text, schemas, operations, and similar OAS-level constructs. Codified by CAMARA Commonalities (the API Design Guide and related artifacts) for use by API Sub Projects. CAMARA-governed.

Reviewer: ```suggestion
- **ICM design info**: information defined by an ICM version that applies to API definitions at API design time — scope format, `securitySchemes` syntax, mandatory `info.description` text, schemas, operations, and similar OAS-level constructs. Supported by guidelines from CAMARA Commonalities (the API Design Guide and related artifacts) for use by API Sub Projects. CAMARA-governed.
```

> - **ICM deployment info**: information defined by an ICM version that applies to deployments at runtime — auth flows, grant types, assertion format and lifetime, token processing, claim handling, and similar runtime behaviors negotiated between an API Provider and API Consumer. API-Provider-governed in the contractual relationship with API Consumers.

Reviewer: ```suggestion
- **ICM deployment info**: information defined by an ICM version that applies to API deployments (including at API Provide/Consumer design time and at runtime) — auth flows, grant types, assertion format and lifetime, token processing, claim handling, and similar behaviors agreed between an API Provider and API Consumer. API-Provider-governed in their contractual relationship with API Consumers.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/71482ed

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/71482ed

> - **API version ICM-compatibility**: the guarantee that an API version's OAS definition respects the ICM design info of a given ICM version. A property of the API definition relative to an ICM version.

Reviewer: ```suggestion
- **API version ICM-compatibility**: the guarantee that an API version's OAS definition respects the ICM design info of a given ICM version. A property of the API definition relative to an ICM version declared in its `x-camara-min-icm` field.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/71482ed

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/71482ed

> - **Deployment ICM-compatibility**: the guarantee that an API Provider's or API Consumer's deployment (a) deploys ICM-compatible API versions and (b) implements the ICM deployment info of the ICM version it claims to operate under.

Reviewer: ```suggestion
- **API deployment ICM-compatibility**: the guarantee that an API Provider's or API Consumer's deployment (a) deploys ICM-compatible API versions and (b) implements the ICM deployment info of the ICM version it claims to operate under.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/71482ed

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/71482ed

> - **ICM-compatibility** (umbrella): both aspects together. Successful integration of an API Consumer with an API Provider requires both to hold.

Reviewer: ```suggestion
- **ICM-compatibility** (umbrella): both API version and API deployment ICM-compatibility together. Successful integration of an API Consumer with an API Provider requires both to hold.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/71482ed

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/71482ed

- **ICM version**: a Semantic Versioning (SemVer 2.0) compliant version number with major, minor and patch components, identifying a specific set of ICM artefacts. Starting with ICM 1.0.0, major-version increments indicate breaking changes for API definitions, API deployments, or both — changes that cannot be expressed additively.
- **ICM lifecycle states**: Supported / Deprecated / Retired / Revoked are the four possible lifecycle states of an ICM version (see §6). Lifecycle states apply per ICM version.
- **ICM governance**: the decision process to transition an ICM version to a different lifecycle state (see §6).
- **Exception (waiver)**: a time-bound, governance-approved authorization that permits a specific (API version, ICM version) pair outside the normal ICM-compatibility guarantee.
- **Compatibility matrix**: the derived artifact listing which (API version, ICM version) pairs are ICM-compatible at a given point in time (see §11).

> ## 3. ICM-compatibility — two aspects, two responsibilities

Reviewer: ```suggestion
## 3. ICM-compatibility
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/71482ed

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/71482ed

> ICM-compatibility has two distinct aspects. Each has its own scope, governance, and signaling mechanism, and each can be impacted by a given ICM change independently of the other.

Reviewer: ```suggestion
ICM-compatibility concerns two distinct sets of information defined by an ICM version referred to as ICM design info and ICM deployment info. 

Each set has its own scope, governance, and signaling mechanism, and each or both can be changed in a given ICM version independently of the other.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/71482ed

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/71482ed

### 3.1 API version ICM-compatibility (design-time)

**An API version is ICM-compatible with an ICM version when its OAS definition respects the ICM design info of that ICM version** — its scope format, `securitySchemes` shape, schemas, operations, and `info.description` text use constructs and conventions defined by the ICM version.

> - **Owned and governed by CAMARA.** ICM design info is codified by the CAMARA Commonalities API Design Guide, which mandates how an API definition must align with ICM. API Sub Projects produce API versions that conform.

Reviewer: ```suggestion
- **Owned and governed by CAMARA.** ICM design info is captured by the CAMARA Commonalities API Design Guide, which mandates how an API definition must align with ICM. API Sub Projects produce API versions that conform to the ICM version by following these guidelines.
```

Author: Applied partially in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/71482ed: kept "codified by" (more accurate than "captured by"); added the trailing clause "by following these guidelines".

> - **Signaled by `x-camara-min-icm`** (§7) declared in the API version's OAS file at API public release time.

Reviewer: ```suggestion
- **Declared in the `x-camara-min-icm`** field (§7) in the API version's OAS definition at API public release time.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/71482ed as "Declared via `x-camara-min-icm`" — keeps the declarative framing while sidestepping "Signaled" (which conflated the property and the field).

- **Maintained in the compatibility matrix** (§11) — published, governed at CAMARA level, and authoritative for which (API version, ICM version) pairs are CAMARA-compliant from the design-time perspective.

> ### 3.2 Deployment ICM-compatibility (runtime)

Reviewer: ```suggestion
### 3.2 API deployment ICM-compatibility (runtime)
```

Author: Applied in subsequent commits. §3.2 heading is "Deployment ICM-compatibility (runtime)" — substantially matches.

**An API Provider's or API Consumer's deployment is ICM-compatible with an ICM version when (a) it deploys ICM-compatible API versions, and (b) the runtime interactions between API Consumer and API Provider implement the ICM deployment info of that ICM version** — auth flows, grant types, assertion format and lifetime, token processing, and claim handling.

- **Owned and governed by the API Provider.** Deployment ICM-compatibility is a contractual matter between API Provider and API Consumer, communicated by the Provider through onboarding documentation, Provider metadata (for example via OIDC discovery), and similar mechanisms. The mandatory `info.description` text in every CAMARA API spec already states that "the specific authorization flows to be used will be agreed upon during the onboarding process."

> - **Discovered at runtime.** An API Consumer determines which ICM version's deployment info applies by inspecting Provider metadata and onboarding artifacts; the API version alone does not pin a specific ICM version on the deployment side.

Reviewer: ```suggestion
- **Identified at deployment/runtime.** An API Consumer determines the applicable ICM version's deployment info by inspecting API Provider metadata and onboarding artifacts; the API version alone does not pin a specific ICM version on the deployment side.
```

Author: Applied in 8a6bbda

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> - **Not recorded in the CAMARA compatibility matrix.** The matrix governs API version ICM-compatibility (design-time). Deployment compliance is the API Provider's responsibility.

Reviewer: ```suggestion
- **Not recorded in the CAMARA compatibility matrix.** The matrix governs API version ICM-compatibility (design-time). API deployment ICM-compatibility is the API Provider's responsibility.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> ### 3.3 The combined view for the API Consumer

Reviewer: ```suggestion
### 3.3 Maintaining ICM-compatibility
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/8a6bbdab6d4f6ec8fb14328e094eccd97890307f

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f with a short lead-in sentence and an actor-duty bullet structure (see reply on r3257512243).

> For successful integration, both aspects must hold: the API version must be ICM-compatible with some ICM version (design-time), and the Provider's deployment must implement a compatible ICM version (runtime). The API Consumer is responsible for ensuring its own implementation matches the deployment ICM version actually in operation.

Reviewer: ```suggestion
To maintain ICM-compatibility of API versions, API designers must release updates of API versions that are no longer ICM-compatible to align with a more recent ICM version

To maintain ICM-compatibility of API deployment, API Providers must ensure that both ICM-compatibility aspects are respected:
- each deployed API version must be ICM-compatible with that ICM version (design-time), or must upgrade to a newer API version. 
- the API Provider must implement an ICM version that allows ICM-compatibility of all deployed APIs (runtime).

Finally, the API Consumer is responsible for ensuring that its own implementation matches the ICM version deployed by the API Provider, as well as the deployed API version.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/8a6bbdab6d4f6ec8fb14328e094eccd97890307f

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f with a short lead-in ("Maintaining ICM-compatibility ... is a joint responsibility of API designers, API Providers, and API Consumers:") and converted to a single bulleted list under that lead. Minor wording tightening: "the ICM version it operates against" rather than orphan "that ICM version", "deployed API versions" rather than "deployed APIs".

> The two aspects can be impacted independently:

Reviewer: ```suggestion
A change in ICM version may impact either or both sets of ICM information independently:
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/8a6bbdab6d4f6ec8fb14328e094eccd97890307f

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> - **An ICM change to design info** (for example, a new scope format) → the API definition may need updating → triggers a new API version for impacted APIs (see §4). API Consumers consuming the new API version must adapt their requests to the new format.

Reviewer: ```suggestion
- **A change to ICM design info** (for example, a new scope format) → the API definition may need updating → triggers a new API version for impacted APIs (see §4). API Consumers consuming the new API version must adapt their requests to the new format.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/8a6bbdab6d4f6ec8fb14328e094eccd97890307f

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> - **An ICM change to deployment info** (for example, tightening assertion lifetime) → API Provider deployments and API Consumer implementations must update → does NOT by itself trigger a new API version. The API definition is unchanged; only the runtime behavior changes.

Reviewer: ```suggestion
- **A change to ICM deployment info** (for example, tightening assertion lifetime) → API Provider deployments and API Consumer implementations must update → does NOT by itself imply a new API version. The API definition is unchanged; only the runtime behavior changes.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/8a6bbdab6d4f6ec8fb14328e094eccd97890307f

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

- **An ICM change to both** → both consequences apply.

> ### 3.4 Path from ICM through Commonalities into API definitions

Reviewer: This section needs to be moved to the end. Once conclusion is reached, it should be checked where/how it best fits.

Author: §3.4 explains how ICM design info reaches API definitions through Commonalities. §6.2's floor formula uses it (the third operand). Moving it to the end disconnects it from where it is used. Suggest keeping it near §6, or folding into §6.1, when we do the final layout.

ICM design info changes do not reach API definitions directly. The path is:

1. ICM defines or updates design info (security schemas, scope format, mandatory text, etc.).
2. Commonalities updates the API Design Guide (and other relevant artifacts) to mandate the new design info for API definitions.
3. API Sub Projects update their API versions to align with the new Commonalities API Design Guide.

This means an ICM design info change typically also triggers a Commonalities update. How tightly ICM and Commonalities lifecycles must be coupled is an open item (see §14).

> ## 4. Minimum ICM version — rule for API version ICM-compatibility

Reviewer: ```suggestion
## 4. Minimum ICM version
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/0ef3907 — §4 content merged into the renamed §6.1 "Minimum ICM version" per your r3260364170. Sections §5..§14 renumbered down to §4..§13.

> ## 4. Minimum ICM version — rule for API version ICM-compatibility

Reviewer: This section 4 needs to be merged into section 7 API version ICM compatibility probably as the first section 7.1

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/0ef3907 — §4 content folded into §6.1 of the renamed §6 ("API version ICM-compatibility - details"). Sections §5..§14 renumbered down to §4..§13.

This section governs the design-time aspect of ICM-compatibility (§3.1). Rules for the deployment-time aspect are in §8.

> An API version declares its minimum ICM version via the `x-camara-min-icm` field in its API definition. This floor expresses the oldest ICM version that the API definition relies on and which is Supported at the time of release of the API version. Subject to ICM's SemVer discipline (§5), the API version is guaranteed ICM-compatible with any ICM version at or above that floor, at least within the same major ICM version.

Reviewer: ```suggestion
An API version declares its minimum ICM version via the `x-camara-min-icm` field in its API definition. This declared ICM version is the oldest ICM version that the API definition relies on and which is Supported at the time of public release of the API version. The API version is guaranteed ICM-compatible with any ICM version at or above this declared version, at least within the same major ICM version.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/8a6bbdab6d4f6ec8fb14328e094eccd97890307f

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> When ICM crosses a major-version boundary, each API definition is assessed by governance at that boundary. API definitions that do not rely on ICM-version-specific constructs may be declared ICM-compatible with the new major through an explicit governance decision recorded in the compatibility matrix. Definitions that do rely on such constructs must be re-released as a new API major targeting the new major ICM version.

Reviewer: ```suggestion
When a new major ICM version is publicly released, the ICM-compatibility of each API version in the ICM-compatibility matrix with this new ICM version needs to be assessed by governance. 
- API versions that do not rely on impacted ICM design info may be declared ICM-compatible with the new major ICM version through an explicit governance decision. This is recorded in the ICM-compatibility matrix. 
- API versions that do rely on the impacted ICM design info release a new major API version with the new major ICM version in their `x-camara-min-icm` field.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/8a6bbdab6d4f6ec8fb14328e094eccd97890307f

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> Example: `x-camara-min-icm: 1.2.0` declared in an API definition means the API depends on ICM 1.x design info from version 1.2.0 onward. The definition is guaranteed ICM-compatible with any Supported or Deprecated ICM version ≥ 1.2.0 in the 1.x major. Whether it remains ICM-compatible with a future 2.x version is a governance assessment made at the time of the ICM 2.0 release.

Reviewer: ```suggestion
Example: `x-camara-min-icm: 1.2.0` declared in an API definition means that the API version depends on ICM v1 design info from version 1.2.0 onward. The definition is guaranteed ICM-compatible with any Supported or Deprecated ICM version ≥ 1.2.0 while in the range of major ICM version v1. Whether an API version remains ICM-compatible with a subsequent ICM version v2 is a governance assessment made at the time of the ICM 2.0.0 public version release.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/8a6bbdab6d4f6ec8fb14328e094eccd97890307f

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> ## 5. ICM SemVer commitment (starting at 1.0.0)

Reviewer: ```suggestion
## 5. ICM versioning
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/8a6bbdab6d4f6ec8fb14328e094eccd97890307f

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> This guideline assumes ICM adopts strict SemVer versioning starting with its 1.0.0 release. The discipline applies to two distinct classes of information changes, mapping to the two aspects of ICM-compatibility (§3):

Reviewer: ```suggestion
This guideline assumes ICM adopts strict SemVer versioning starting with its 1.0.0 release. 

ICM versioning relates to the changes made to the two classes of ICM information that impact ICM-compatibility: APIs design info and API deployment info (§3):
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/8a6bbdab6d4f6ec8fb14328e094eccd97890307f

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> - changes that break **API version ICM-compatibility** — affecting ICM design info (scope format, `securitySchemes` syntax, schemas, operations) and requiring API definitions to be updated;

Reviewer: ```suggestion
- changes that break **API version ICM-compatibility** — affecting ICM design info (scope format, `securitySchemes` syntax, schemas, operations), requiring API versions to be updated;
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> - changes that break **deployment ICM-compatibility** — affecting ICM deployment info (auth flows, grant types, assertion format and lifetime, token processing) and requiring Provider and Consumer implementations to be updated, independently of any API definition.

Reviewer: ```suggestion
- changes that break **API deployment ICM-compatibility** — affecting ICM deployment info (auth flows, grant types, assertion format and lifetime, token processing), requiring API Provider and Consumer implementations to be updated, independently of the API version.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> Both classes are governed by the same SemVer rules:

Reviewer: ```suggestion
Both types of changes are captured in the ICM version according to standard SemVer rules:
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> - **Minor versions (1.x → 1.y)**: additive only. No breaking change to ICM design info (preserves API version ICM-compatibility) and no breaking change to ICM deployment info (preserves deployment ICM-compatibility). New optional flows, new optional claims, new recommendations permitted. Removals, replacements, and tightenings that reject previously-legal API definitions or deployment behavior may only appear at a major-version boundary.

Reviewer: put "major" bullet first

Author: Applied as part of combined §5 bullet reorder (major → minor → patch) + wording updates in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> - **Minor versions (1.x → 1.y)**: additive only. No breaking change to ICM design info (preserves API version ICM-compatibility) and no breaking change to ICM deployment info (preserves deployment ICM-compatibility). New optional flows, new optional claims, new recommendations permitted. Removals, replacements, and tightenings that reject previously-legal API definitions or deployment behavior may only appear at a major-version boundary.

Reviewer: ```suggestion
- **Minor version change (1.x.y → 1.x+1.0)**: additive only. No breaking change, but minor change to ICM design info (preserves API version ICM-compatibility), or no breaking change, but minor change to ICM deployment info (preserves API deployment ICM-compatibility). For example, new optional flows, new optional claims, new recommendations permitted. Minor changes must not reject previously ICM-compatible API deployment behavior, as these may only apply in a major ICM version change.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> - **Major versions (1.x → 2.0)**: reserved for breaking changes that cannot be expressed additively. These may affect ICM design info (for example, replacement of a mandatory security schema, an incompatible scope format, removal of a required claim), affect ICM deployment info (for example, mandatory new authentication rules, an assertion lifetime cap that rejects existing Consumers), or both. Majors should be deliberately rare and driven by security or regulatory necessity.

Reviewer: ```suggestion
- **Major version (1.x.y → 2.0.0)**: reserved for breaking changes that cannot be expressed additively. These may be due to ICM design info changes (for example, replacement of a mandatory security schema, an incompatible scope format, removal of a required claim), to ICM deployment info changes (for example, mandatory new authentication rules, an assertion lifetime cap that rejects existing Consumers), or both. Major ICM versions should be deliberately rare and driven by security or regulatory necessity.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> - **Patch versions (1.2.3 → 1.2.4)**: editorial or defect corrections with no impact on either aspect.

Reviewer: ```suggestion
- **Patch version (1.2.3 → 1.2.4)**: editorial or defect corrections with no impact on either API design info or API deployment info.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> An ICM change that invalidates existing deployment implementations is a major-version event even if no API definition requires re-release. Semver discipline is not satisfied by preserving API version ICM-compatibility alone.

Reviewer: ```suggestion
An ICM change to API deployment info that invalidates existing API deployments requires a major ICM version release, even if no new API versions are required. ICM versioning is not focused only on preserving API version ICM-compatibility.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> The transition to 1.0.0 should coincide with a scope-baseline review of the ICM documents, declaring the then-current definitions as the stable starting point. Pre-1.0 versions are handled by the Legacy section (§12).

Reviewer: this comment should go to an "ICM version governance" section. it is a project issue, not part of the general process description.

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/5bca022 — folded into §12 item 2 (Open governance decisions: "Transition to ICM 1.0.0"). Happy to relocate if a different governance section fits better.

> ## 6. ICM lifecycle states and governance

Reviewer: ```suggestion
## 6. ICM version lifecycle states and governance
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> ### 6.1 State model

Reviewer: ```suggestion
### 6.1 ICM version lifecycle states - definitions
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f as part of a §6 renumber: §6.1 (definitions) + §6.2 (transitions, per your r3261122190) + §6.3 (governance parameters, per your r3258615336) + §6.4 (publication of state) + §6.5 (release-note tables).

> Each ICM version is in exactly one lifecycle state at any given time. States apply per ICM version, with the major ICM version as the primary grain: a new major enters the Supported state at its public release, and sub-major versions (minor and patch) inherit their major's state by default. Governance MAY explicitly transition a specific sub-major version to a different state.

Reviewer: ```suggestion
Each ICM version is in exactly one lifecycle state at any given time. States apply per ICM version. A  new major ICM version enters the Supported state at its public release. Subsequent minor or patch versions inherit the state of their major ICM version by default. Governance MAY explicitly transition a specific minor or patch version to a different state.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> The table below summarizes each state's meaning for the ICM version itself and its consequences for API versions and API deployments (mapping the two-axis decomposition of §3 onto the state model).

Reviewer: ```suggestion
The table below summarizes the meaning of each state for the ICM version itself, and its consequences for API versions and API deployments.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

| State | ICM version | API version | API deployment |
|---|---|---|---|
| **Supported** | OK to use | may be used as `x-camara-min-icm` | may operate against this ICM version |

> | **Deprecated** | Sunset announced; migration window active | SHOULD NOT be used as `x-camara-min-icm` | may continue to operate against this ICM version |

Reviewer: ```suggestion
| **Deprecated** | Sunset announced (planned end of life); migration window active | SHOULD NOT be used as `x-camara-min-icm` | may continue to operate against this ICM version |
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> | **Retired** | Terminal (planned end of life) | is no longer ICM-compatible; MUST evolve to a Supported ICM version or be Retired at the API level | MUST migrate to an ICM-compatible deployment with a Supported ICM version |

Reviewer: ```suggestion
| **Retired** | Terminal (end of life) | is no longer ICM-compatible; MUST evolve to a Supported ICM version or be Retired at the API level | is no longer ICM-compatible; MUST migrate to an ICM-compatible deployment with a Supported ICM version |
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> | **Revoked** | Terminal (exceptional); replaced by a later minor or patch version due to defects or security issues and MUST NOT be used; governance names the replacement version | MUST re-establish ICM-compatibility with a Supported ICM version | MUST migrate to a Supported ICM version and ICM-compatible deployment |

Reviewer: ```suggestion
| **Revoked** | Terminal (exceptional end-of-life); replaced by a later minor or patch version due to defects or security issues and MUST NOT be used; governance names the replacement version | MUST re-establish ICM-compatibility with a Supported ICM version | MUST migrate to a Supported ICM version and ICM-compatible deployment |
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> | **Revoked** | Terminal (exceptional); replaced by a later minor or patch version due to defects or security issues and MUST NOT be used; governance names the replacement version | MUST re-establish ICM-compatibility with a Supported ICM version | MUST migrate to a Supported ICM version and ICM-compatible deployment |

Reviewer: "replaced by a later minor or patch version due to defects or security issues": could it not also be replaced by (rolled back to) a previous minor or patch version if the current one introduced an issue ?

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/5bca022

> Note: the "API version" and "API deployment" columns describe consequences within CAMARA governance scope — i.e., for API versions published in the CAMARA catalog and for CAMARA-compliant API Provider deployments. Use outside that scope is not governed by this guideline.

Reviewer: ```suggestion
Note: the "API version" and "API deployment" columns describe impacts (in CAMARA governance scope) on ICM-compatibility of API versions as declared in the ICM-compatibility matrix and on ICM-compatibility of API deployments. Use outside that scope is not governed by this guideline.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> 

Reviewer: ```suggestion

### 6.2 ICM version lifecycle states - transitions

```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f as new §6.2 — transition-related paragraphs (planned sequence, per-version transitions, Notes about Retired alignment and Deprecation cascade) moved from §6.1 into the new section. Subsequent subsections renumbered to §6.3 / §6.4 / §6.5.

> The planned state-transition sequence is Supported → Deprecated → Retired. Revoked is entered as an exceptional action outside this sequence and requires an explicit replacement version.

Reviewer: ```suggestion
The planned (default) state-transition sequence of an ICM version is 

Supported → Deprecated → Retired

The Revoked state is entered through an exceptional transition decided by governance outside the planned sequence and requires an explicit replacement version to be identified.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> For Supported major ICM versions, governance MAY transition specific minor or patch versions to Deprecated, Retired, or Revoked when they should no longer be used as a floor — for example, when known ambiguities or defects are resolved in a later minor version, or when a critical defect requires replacing a specific patch version. Such per-version transitions do not impact ICM-compatibility of API versions or API deployments — later minor ICM versions remain compatible with earlier ones by SemVer — but they influence the "lowest Supported minor" used in §7.2 and are relevant for API deployments.

Reviewer: ```suggestion
For Supported major ICM versions, governance MAY transition specific minor or patch ICM versions to Deprecated, Retired, or Revoked when they should no longer be used. 
- For example, when known ambiguities or defects are resolved in a later minor ICM version, or when a critical defect requires replacing a specific patch ICM version. Such per-version transitions do not impact ICM-compatibility of API versions or API deployments. 
- Later minor ICM versions remain compatible with earlier ones by SemVer, but they influence the "lowest" Supported minor ICM version used in §7.2 and are relevant for API deployments.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> Note: the term "Retired" aligns with the API lifecycle terminology proposed in [ReleaseManagement#459](https://github.com/camaraproject/ReleaseManagement/issues/459), so that ICM and API lifecycles use the same vocabulary for the terminal state.

Reviewer: ```suggestion
Note: the term "Retired" aligns with the API lifecycle terminology, so that ICM and API lifecycles use the same vocabulary for the terminal state.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> Note: Deprecation or Retirement of an ICM major does not by itself Deprecate or Retire API versions referencing it — only the corresponding cells in the compatibility matrix change. API version lifecycle (Deprecation, Retirement at the API level) is governed independently by CAMARA's API lifecycle process (see [ReleaseManagement#459](https://github.com/camaraproject/ReleaseManagement/issues/459)).

Reviewer: ```suggestion
Note: Deprecation or Retirement of a major ICM version does not by itself Deprecate or Retire the API versions that reference it in their `x-camara-min-icm` field. It only changes the corresponding entries in the ICM-compatibility matrix. API version lifecycle (Deprecation, Retirement at the API level) is governed independently by CAMARA's API lifecycle process (see tbd [API lifecycle states](https://github.com/camaraproject/ReleaseManagement/issues/459)).
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> ### 6.2 Governance parameters

Reviewer: ```suggestion
### 6.2 Governance parameters for ICM lifecycle states
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f as **§6.3** (your suggestion labelled it §6.2; with the new §6.2 transitions inserted per r3261122190, the governance-parameters subsection lands at §6.3 in the renumbered scheme).

| Parameter | Suggested starting value | Notes |
|---|---|---|

> | Duration of Supported state for an ICM major after its successor is published | ≥ 18 months | After a newer major ICM version is published, the previous major remains Supported for at least this period before governance transitions it to Deprecated. During this period, API Providers plan to migrate to the newer ICM major. |

Reviewer: ```suggestion
| Duration of Supported state (of a major ICM version after its successor major ICM version is published) | ≥ 18 months | After a newer major ICM version is published, the previous major remains Supported for at least this period before governance transitions it to Deprecated. During this period, API Providers are expected to plan migration to the newer major ICM version. |
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> | Duration of Deprecated state | ≥ 12 months | Active migration period for deployments before the major is Retired. |

Reviewer: ```suggestion
| Duration of Deprecated state | ≥ 12 months | Active migration period for API deployments before the major ICM version is Retired and impacts the API deployment's ICM-compatibility |
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> | Concurrent support requirement | Latest published Supported major ICM version + most recent previous Supported major (during overlap) | Applies to CAMARA-compliant API Providers during this period. |

Reviewer: ```suggestion
| Concurrent support requirement by API deployments | API Providers shall continue to deploy the most recent previous Supported major ICM version next to the latest published Supported major ICM version while this previous ICM major version is still Supported (see period defined above) | Applies to ICM-compatible API deployments during this period. |
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> | Security kill-switch | Conditions permitting acceleration of any state transition | Explicit governance action per incident; see §13. |

Reviewer: ```suggestion
| Security exception | Conditions permitting acceleration of any state transition | Explicit governance action per incident; see §13. |
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> These values are starting points for WG discussion.

Reviewer: ```suggestion
```

Author: Annotated rather than dropped: kept the sentence with an HTML comment `<!-- to be removed when WG agrees -->` next to it, applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/5bca022. The sentence will come out cleanly when the WG signs off on the parameter values.

> ### 6.3 Publication of state

Reviewer: ```suggestion
### 6.3 ICM version - Release notes

#### 6.3.1 Publication of life cycle state

```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/b34308d99bff567807c5cd01b046879d9b3d9ed7.

> ### 6.3 Publication of state

Reviewer: I assume the lifecycle state table and the change tables shall be documented in the CHANGELOG.md file of ICM repo, correct ? (not in the release notes of the GitHub release). If so this should be made more explicit in the text.

Author: Resolved without a doc change: CHANGELOG.md and the GitHub release body carry the same content (the CHANGELOG entry for a version is copied into the release body). The generic "release notes" term in §5.4 covers both, so no special wording is needed.

> State is published in each ICM release's notes, as a standard table in the release notes template. No separate governance artifact is required — each ICM release carries the current state for all active ICM majors. Between ICM releases, the state from the latest release is authoritative. State transitions are committed at ICM release time unless an out-of-cycle governance action specifies otherwise.

Reviewer: ```suggestion
The lifecycle state of an ICM version is published in each ICM version's release notes, as a standard table in the release notes template. No separate governance artifact is required. Each ICM release carries the current state for all major ICM versions. State transitions are committed at ICM release time unless an out-of-cycle governance action specifies otherwise.
```

Author: Superseded by your later suggestion at https://github.com/camaraproject/ReleaseManagement/pull/497#discussion_r3262052803; later version applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> State is published in each ICM release's notes, as a standard table in the release notes template. No separate governance artifact is required — each ICM release carries the current state for all active ICM majors. Between ICM releases, the state from the latest release is authoritative. State transitions are committed at ICM release time unless an out-of-cycle governance action specifies otherwise.

Reviewer: ```suggestion
The lifecycle state is published in each ICM version's release notes, as a table in the release notes template. No separate governance artifact is required. Each ICM version release carries the lifecycle state for all ICM versions. State transitions are committed at ICM public release unless an out-of-cycle governance action specifies otherwise.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> A machine-readable schema for the published state may be defined later to support automated tooling, once the state-model concepts in this section are settled. Until then, the ICM release notes are the single authoritative source.

Reviewer: ```suggestion
A machine-readable schema for the published lifecycle state of ICM versions may be defined later to support automated tooling. Until then, the ICM version release notes are the single authoritative source.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> ### 6.4 ICM release notes — design info and deployment info change tables

Reviewer: ```suggestion
#### 6.3.2 ICM version change tables
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/b34308d99bff567807c5cd01b046879d9b3d9ed7.

> In addition to the state table, each ICM release SHOULD document **breaking changes** introduced in that ICM version, split by the two aspects of ICM-compatibility (§3). Non-breaking changes (additive features, clarifications) are documented in the regular CHANGELOG and do not require entries in the tables below.

Reviewer: ```suggestion
In addition to the lifecycle state table, each ICM release MUST document **breaking changes** introduced in that ICM version, split by ICM design info changes and ICM deployment info changes, covering the two aspects of ICM-compatibility (§3). Non-breaking changes (additive features, clarifications) are documented in the regular CHANGELOG and do not require entries in the tables below.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

**Table A — Breaking ICM design info changes** (impacting API definitions; see §3.1). Used by API Sub Projects to assess whether their API versions need a new release.

| ICM version | Breaking design info change | Affected construct | Impact on API definitions |
|---|---|---|---|
| _example to fill in: e.g., "1.x → 2.0 introduces new scope format X"_ | _what changed_ | _where it appears in an API definition_ | _what API teams need to do_ |

> **Table B — Breaking ICM deployment info changes** (impacting runtime deployments; see §3.2). Used by API Providers and API Consumers to plan deployment updates.

Reviewer: ```suggestion
**Table B — Breaking ICM deployment info changes** (impacting API deployments; see §3.2). Used by API Providers and API Consumers to plan deployment updates.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

| ICM version | Breaking deployment info change | Affected runtime behavior | Impact on deployments |
|---|---|---|---|
| _example to fill in: e.g., "0.3.0 introduces 300s client-assertion lifetime cap"_ | _what changed_ | _which deployment behavior_ | _what Provider and Consumer need to do_ |

The exact column layout is subject to refinement once concrete examples are filled in (see §14).

> ## 7. API spec declaration: `x-camara-min-icm`

Reviewer: ```suggestion
## 7. API version ICM-compatibility - details```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/0ef3907 — landed at **§6** (not §7) after the §4→§6.1 merge and renumber.

> ### 7.1 Declaration

Reviewer: ```suggestion
### 7.1 Minimum ICM version
```

Reviewer: merge section 4 into 7.1 here

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/0ef3907 — landed at **§6.1** after the §4→§6.1 merge and renumber.

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/0ef3907 — §4 content folded into §6.1 (was §7.1). Section numbering shifted down by one after §4 was removed.

> Every CAMARA API definition declares a single top-level extension field naming the minimum required version of the CAMARA Identity and Consent Management specification:

Reviewer: ```suggestion
Every CAMARA API version declares in its API definition a top-level extension field `x-camara-min-icm` identifying the minimum required version of the CAMARA Identity and Consent Management specification:
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

```yaml
info:
  # Minimum version of the CAMARA Identity and Consent Management (ICM)
  # specification this API definition is built against.
  x-camara-min-icm: 1.2.0
```

- Three-part semantic version matching the ICM release versioning scheme.
- Names the lowest ICM version against which the API definition is ICM-compatible.
- Names the major ICM version family this API is initially compatible with (e.g., `1.2.0` → ICM 1.x). Cross-major compatibility may be extended via §4 governance assessment.

> - Publication-time fixed. Does not change after the API version is released.

Reviewer: ```suggestion
- Identifies the lowest ICM version against which the API definition is ICM-compatible using the full SemVer version of ICM (x.y.z).
- Is defined at API public release. Does not change after the API version is released.
- Possibly includes an extension to allow covering compatibility across multiple major ICM versions (later)
```

Author: The field was dropped in the May 13 working merge — cross-major compatibility is now handled by governance assessment recorded in the matrix (§6.1), not by a spec extension. Adding it back would reverse that. If we want to revisit, better as a §12 open item.

> - The field name follows the CAMARA convention for ICM-related extensions (parallel to `x-camara-commonalities`), with no `-version` suffix; the field name itself denotes a version.

Reviewer: ```suggestion
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> ### 7.2 Rule for choosing the value at release

Reviewer: ```suggestion
### 7.2 Setting the `x-camara-min-icm` value
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> At each API release (new major, minor, or patch), the API Sub Project sets the value as:

Reviewer: ```suggestion
At each API version public release (new major, minor, or patch), the API Sub Project sets the value as:
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

```

> x-camara-min-icm := max(

Reviewer: ```suggestion
x-camara-min-icm = max (
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

  lowest Supported ICM version at release time,
  lowest ICM version containing all features this API's definition requires,
  lowest ICM version required by the Commonalities version declared in `x-camara-commonalities`
)
```

> For APIs with no ICM-version-specific feature dependencies beyond what Commonalities mandates, the second operand collapses to zero and the rule reduces to `max(lowest Supported ICM version at release time, ICM floor required by the declared Commonalities version)`. Publishing a new API version therefore always raises its floor to a Supported ICM major. An API Provider cannot declare a newly published API version as CAMARA compliant while offering only a Deprecated or Retired ICM major, even if the definition would technically work with those older versions.

Reviewer: ```suggestion
For APIs with no ICM-version-specific feature dependencies beyond what Commonalities mandates, the second element is not applicable which reduces the rule to `max (lowest Supported ICM version at API version public release, lowest ICM version required by the Commonalities version declared in `x-camara-commonalities`). 
Publishing a new version of such an API therefore always raises its `x-camara-min-icm` version to the latest Supported major(?) ICM version. _(Note tdg: TBC not fully sure this is correct)_

An API Provider cannot declare a newly deployed API version as ICM-compatible while offering only a Deprecated or Retired major ICM version, even if the API version would technically work with those older ICM versions.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> Past API versions retain their published value; the floor is publication-time fixed even if the lowest Supported ICM version moves upward over time. Such declarations remain functionally valid within their original ICM major; cross-major compatibility is assessed by governance per §4.

Reviewer: ```suggestion
Past API versions retain their published `x-camara-min-icm` value (fixed at API public release) even if the lowest Supported ICM version moves upward over time. The API version remain ICM-compatible by default with any minor ot patch updates of the ICM version. ICM-compatibility across different major ICM versions is assessed by governance per §7.1(?).
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> ### 7.3 Validation

Reviewer: ```suggestion
### 7.3 CAMARA validation support
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> A CAMARA linting rule, run at API release time, reads the current ICM state (from the latest ICM release's notes) and verifies:

Reviewer: ```suggestion
A CAMARA linting rule runs at API version release and gets the lifecycle state of the API declared ICM version from the latest ICM version's release's notes. The rule then verifies:
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/b996275 — added an explicit dependency note in §6.3 pointing at §5.4.1, and tightened §5.4.1 from "may be defined later" to "must be available" for the validation rule to consume.

> 1. `x-camara-min-icm` is present.

Reviewer: ```suggestion
1. `x-camara-min-icm` is present in the API definition.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

2. Its value is a syntactically valid SemVer 2.0 version.

> 3. The value is a Supported ICM version at release time (the feature-dependency floor and the Commonalities-required floor per §7.2 may raise the minimum).

Reviewer: ```suggestion
3. The value refers to a Supported ICM version at API version public release.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> ## 8. API Provider responsibilities — deployment ICM-compatibility

Reviewer: ```suggestion
## 8. API deployment ICM-compatibility - details
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> This section governs the deployment-time aspect of ICM-compatibility (§3.2). Rules for the design-time aspect (API version ICM-compatibility) are in §4.

Reviewer: ```suggestion
This section describes the governance of API deployment ICM-compatibility (runtime).
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

An API definition might technically work against an ICM version that the compatibility matrix (§11) does not record as CAMARA-compliant — for example, a Retired or Revoked ICM version still running somewhere. Such pairings are outside CAMARA governance and cannot be claimed as CAMARA compliant.

> CAMARA-compliant API Providers MUST:

Reviewer: ```suggestion
### 8.1 API Provider responsibilities

To achieve ICM-compatibility of their API deployment, API Providers MUST:
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> 1. Implement the latest published Supported major ICM version.

Reviewer: ```suggestion
1. Implement a Supported ICM version.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> 2. During the period in which a previous ICM major remains in the Supported state, also implement that major, so that API versions targeting either Supported major can be served.

Reviewer: ```suggestion
2. When a new major ICM version is published and is in Supported state, plan implementation of that ICM version. During the period in which the previous major ICM version remains in the Supported state, keep that major ICM version running in parallel, so that API versions targeting either Supported ICM version can be deployed.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> 3. Publish the ICM versions they implement, in a form that API Consumers can discover (for example via Provider metadata, onboarding documentation, or the standard `info.description` template mandated by Commonalities).

Reviewer: ```suggestion
3. Make the information about the deployed ICM versions available to the API Consumers (for example via Provider metadata, onboarding documentation, or the standard `info.description` template mandated by Commonalities).
```

Author: Not applied literally — your follow-up at https://github.com/camaraproject/ReleaseManagement/pull/497#discussion_r3262528245 ("Not convinced by 'or the standard info.description template mandated by Commonalities' — suggest to remove that") was applied instead. The info.description template does not advertise the deployed ICM version, so the clause has been dropped from §7.1 item 3 in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/5bca022.

> 3. Publish the ICM versions they implement, in a form that API Consumers can discover (for example via Provider metadata, onboarding documentation, or the standard `info.description` template mandated by Commonalities).

Reviewer: Not convinced by the part "or the standard `info.description` template mandated by Commonalities)." Suggest to remove that.

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/5bca022

> 4. Publish **two separate compliance statements**:

Reviewer: ```suggestion
4. Publish **two separate ICM-compatibility statements**:
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

   - **API-specification compliance**: for each API version offered (the API-version-ICM-compatibility side, §4).
   - **ICM-implementation compliance**: which major ICM versions the Provider implements (the deployment-ICM-compatibility side).

These two statements together constitute the API Provider's CAMARA-compliance posture.

> 

Reviewer: ```suggestion
   - **API version ICM-compatibility**: for each API version offered.
   - **API deployment ICM-compatibility**: the major ICM versions implemented.

These two statements together constitute the API Provider's ICM-compatibility declaration.

```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> The mechanism by which an API Provider implements multiple ICM majors concurrently is an implementation choice and is not prescribed by this guideline; what matters is that each implemented major ICM version is discoverable and usable by API Consumers during the relevant period.

Reviewer: ```suggestion
The mechanism by which an API Provider implements multiple ICM major versions concurrently is an implementation choice and is not prescribed by this guideline; what matters is that each implemented ICM version is clearly announced by the API Provider for use by API Consumers during the relevant migration period.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> ### 8.1 API Consumer responsibilities

Reviewer: ```suggestion
### 8.2 API Consumer responsibilities
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> Successful end-to-end integration requires the API Consumer to implement its side of the ICM deployment info — auth flows, grant types, assertion format and lifetime, token processing — consistently with the ICM version implemented by the API Provider it interacts with. The API Consumer determines this ICM version through Provider metadata and onboarding artifacts, not through the API version alone. API Consumers are not bound by this guideline directly; this section describes what an API Consumer needs to do to integrate with a CAMARA-compliant API Provider.

Reviewer: ```suggestion
Successful API deployments requires the API Consumer to implement its side of the ICM deployment info — auth flows, grant types, assertion format and lifetime, token processing — consistently with the ICM version implemented by the API Provider it interacts with. 

The API Consumer determines the applicable ICM version through Provider metadata and onboarding artifacts, not through the API version alone. API Consumers are not bound by this guideline directly.

The API Consumer is responsible for ensuring that its own implementation matches both the  the announced ICM version and the API version.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> ## 9. Aggregator role

Reviewer: I am not sure we need this section. There is no guidelines given in the section itself, so it could be dropped.

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/5bca022

> Aggregators accept suppliers offering (API version, ICM version) pairs across a spectrum and present API Consumers with a narrower, consistent set of offerings — typically two or three API versions and one or two ICM majors, giving API Consumers time to migrate on both axes.

Reviewer: ```suggestion
Aggregators accept a variety of supplier offering pairs (API version, ICM version) and present API Consumers with a narrower, consistent set of API offerings — typically two or three API versions and one or two major ICM versions. This approach gives API Consumers time to migrate with respect to both ICM version changes and API version changes.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

Aggregators operate as ecosystem normalizers:

> - They MAY wrap or transform between ICM majors (for example, accepting ICM 2.x assertions from API Consumers while calling ICM 1.x suppliers, or vice versa).

Reviewer: ```suggestion
- They MAY wrap or transform between major ICM versions (for example, accepting ICM v2 assertions from API Consumers while calling ICM v1 API Providers, or vice versa).
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> - They MAY narrow the supplier pool to those operating at specific ICM majors they choose to offer.

Reviewer: ```suggestion
- They MAY narrow the set of API Providers to those operating with the specific major ICM versions the aggregator chooses to offer to API Consumers.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> - They MAY encourage convergence among their suppliers toward a common set of (API, ICM) pairs by aligning their commercial onboarding and support processes accordingly.

Reviewer: ```suggestion
- They MAY encourage convergence among their API Providers toward a common set of (API version, ICM version) pairs by aligning their commercial onboarding and support processes accordingly.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> This guideline recognizes aggregators as a first-class actor in the CAMARA ecosystem. The matrix of (API version, ICM version) pairs across raw suppliers is a consequence of API and ICM versions evolving on independent cadences; aggregators handle matrix reduction for their API Consumers, and CAMARA governance provides the outer envelope within which both suppliers and aggregators operate.

Reviewer: ```suggestion
This guideline recognizes aggregators as a first-class actor in the CAMARA ecosystem. The ICM-compatibility matrix of (API version, ICM version) pairs across suppliers is a consequence of API and ICM versions evolving on independent cadences; aggregators handle matrix reduction for their API Consumers, and CAMARA governance provides the baseline on which both API Providers and aggregators operate.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> ## 10. Release governance cadence

Reviewer: ```suggestion
## 10. ICM version release cadence
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

### 10.1 Signal / Sync alignment

> ICM releases are announced primarily at Signal meta-releases (H1 each year); API releases primarily at Sync meta-releases (H2). This natural cadence provides lead time:

Reviewer: ```suggestion
The Signal meta-release is used for 
- planning and releasing a new ICM version as needed
- deciding and recording previous ICM version lifecycle transitions (lifecycle state table)

New ICM versions MAY be introduced outside Signal meta-releases (out-of-cycle) if required.

ICM releases are done at Signal meta-release (first half of each year); API releases primarily at Sync meta-release (second half of each year). This cadence provides lead time as follows:
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> 1. **Signal Year N**: ICM version published with state table. API Providers plan to move to the newer ICM version. API Sub Projects plan their new API versions.

Reviewer: ```suggestion
1. **Signal Year N**: ICM version published with updated ICM version lifecycle state table. API Sub Projects plan a new API version as needed for Sync Year N. API Providers plan to move their ICM implementation to the newer ICM version.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> 2. **Signal → Sync gap (~6 months)**: API Providers move their ICM implementation forward; API Sub Projects produce new API versions.

Reviewer: ```suggestion
2. **Signal → Sync (~6 months)**: API Providers move their ICM implementation forward; API Sub Projects produce new API versions.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> 3. **Sync Year N**: new API versions land, declaring in `x-camara-min-icm` the newly Supported major ICM version if they use new features it introduces.

Reviewer: ```suggestion
3. **Sync Year N**: new API versions are released, declaring in their `x-camara-min-icm` the newly Supported ICM version if the API version uses new features that the ICM version introduces.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

### 10.2 Out-of-cycle ICM releases

> Out-of-cycle ICM releases are permitted and sometimes required — e.g., for security patches, defect corrections, or urgent regulatory changes. These follow the same SemVer discipline. State transitions may occur off-cycle in security-driven cases, with explicit governance record.

Reviewer: ```suggestion
Out-of-cycle ICM releases are allowed and sometimes required, e.g., for security patches, defect corrections, or urgent regulatory changes. These changes impact the ICM version as per SemVer guidelines. 

ICM version lifecycle state transitions may occur off-cycle in security-driven cases, or on explicit governance decision.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> ### 10.3 Signal is not a gate

Reviewer: Propose to drop this section. I moved part of the text to section 10.1.

Author: Section dropped in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f; out-of-cycle content folded into the new §10.1 and §10.2 wording per your earlier suggestions.

Signal timing is an announcement and coordination anchor, not a mandatory gate. ICM versions MAY be introduced outside Signal when necessary. API versions MAY be released outside Sync. Governance focuses on content discipline (SemVer correctness, state assignments) rather than calendar alignment.

> ## 11. Compatibility matrix as derived artifact

Reviewer: ```suggestion
## 11. ICM-compatibility matrix
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> The API–ICM compatibility matrix is maintained as a **derived artifact**, computed automatically from:

Reviewer: ```suggestion
The ICM-compatibility matrix between API versions and ICM versions is maintained as a **derived artifact**, computed automatically from:
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> - API version declarations (`x-camara-min-icm` in each published API definition).

Reviewer: ```suggestion
- API definition declarations (`x-camara-min-icm` in each published API version).
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> - ICM state per published ICM version (from the latest ICM release notes), including any per-version overrides.

Reviewer: ```suggestion
- ICM version lifecycle state table (from the latest ICM version release notes), including any per-version overrides.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> - Exception waivers (if any).

Reviewer: ```suggestion
- Exceptions (if any).
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> - Governance decisions extending an API definition's ICM-compatibility across an ICM major boundary (per §4).

Reviewer: ```suggestion
- Governance decisions may exceptionally extend an API version's ICM-compatibility to cover an additional major ICM version.
- Governance decisions may exceptionally extend the duration of an ICM version's Supported or Deprecated lifecycle state. This extends the duration of API version ICM compatibility accordingly.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> The matrix lists one row per published API version and one column per published ICM version. It records which (API version, ICM version) pairs are CAMARA-compliant.

Reviewer: ```suggestion
The matrix lists one row per released public API version and one column per released public ICM version. It records which pairs (API version, ICM version) are ICM-compatible.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> ### 11.1 Compatibility computation

Reviewer: ```suggestion
### 11.1 ICM-compatibility matrix derivation
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

```
compatible(API vX, ICM version vY) :=
  vY's major == major(API vX's x-camara-min-icm)
  AND vY >= API vX's x-camara-min-icm
  AND vY is in state {Supported, Deprecated}
  OR a governance decision extends the API's ICM-compatibility to vY's major

>   OR an approved, time-bound compatibility waiver exists

Reviewer: ```suggestion
ICM-compatibility (API vX, ICM vY) =
  ICM vY.major == API vX x-camara-min-icm.major)
  AND ICM vY >= API vX x-camara-min-icm
  AND ICM vY is in state {Supported, Deprecated}
  OR a governance decision extends the API vX ICM-compatibility to vY.major
  OR a governance approved, time-bound ICM-compatibility exception exists
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

```

> The state condition applies to the specific ICM version `vY`, inheriting from its major unless a per-version override is in place. Versions in the Retired or Revoked state are excluded from all positive compatibility cells, except via an approved waiver.

Reviewer: ```suggestion
The lifecycle state condition applies to the specific ICM version `vY`, inheriting from its major ICM version unless a per-version override is in place. ICM versions in the Retired or Revoked state are excluded from the ICM-compatibility matrix, unless an approved exception has been decided.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> Cells that are "compatible" represent legitimate (API, ICM) pairs that API Providers may offer and API Consumers may consume.

Reviewer: ```suggestion
The matrix of ICM-compatible pairs (API version, ICM version) are the combinations that API Providers may offer and API Consumers may consume.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> ### 11.2 Ownership

Reviewer: ```suggestion
### 11.2 ICM-compatibility matrix - ownership
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> The matrix is published by Release Management. It is computed, not hand-edited. Only exception waivers require human governance action.

Reviewer: ```suggestion
The ICM-compatibility matrix is published by Release Management. It is computed, not hand-edited. Only exceptions require human governance action.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> ### 11.3 Scope

Reviewer: ```suggestion
### 11.3 ICM-compatibility matrix - updates
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> Matrix update triggers are:

Reviewer: ```suggestion
ICM-compatibility matrix updates are triggered by:
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> - New ICM release (state may change).

Reviewer: ```suggestion
- New ICM version public release (lifecycle state of other ICM versions may change).
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> - New API release (new row added; old rows unchanged).

Reviewer: ```suggestion
- New API version public release (new row added to the matrix; existing rows unchanged).
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

- Exception granted, modified, or expired.

## 12. Legacy ICM 0.x handling

> This guideline's SemVer-based rules take effect with ICM 1.0.0. For the pre-1.0 transition period, Release Management publishes a one-time historical record documenting which ICM 0.x transitions were client-facing-breaking:

Reviewer: ```suggestion
This guideline's SemVer-based rules take effect with starting with ICM version 1.0.0. For the pre-v1.0.0 versions, Release Management publishes a one-time historical record documenting which ICM v0.x.y versions introduced client-facing breaking changes:
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> | ICM transition | Client-facing forward break? | Source PRs |

Reviewer: ```suggestion
| ICM version | Lifecycle state | Client-facing forward break? | Source PRs |
```

Author: Not included ... we should not mix the guideline with the publication of lifecycle states (they will change over time) -> mechanism for that is already documented above (release notes of ICM).

> |---|---|---|

Reviewer: ```suggestion
|---|---|---|---|
```

Author: not included see above

| 0.2.0 → 0.3.0 | Yes — client-assertion lifetime cap | [#216](https://github.com/camaraproject/IdentityAndConsentManagement/pull/216) |
| 0.3.0 → 0.4.0 | Yes — mandatory signed-request fields; error-code rename | [#285](https://github.com/camaraproject/IdentityAndConsentManagement/pull/285), [#287](https://github.com/camaraproject/IdentityAndConsentManagement/pull/287) |
| 0.4.0 → 0.5.0 | No — additive | — |

> State for 0.x versions is assigned by ICM governance as a one-time exercise. The full SemVer discipline does not apply retroactively to pre-1.0 versions.

Reviewer: ```suggestion
| 0.6.0 | Supported | ?, xxx | — |
| 0.5.0 | Supported | No, additive only | — |
| 0.4.0 | Supported | Yes, mandatory signed-request fields; error-code rename | [#285](https://github.com/camaraproject/IdentityAndConsentManagement/pull/285), [#287](https://github.com/camaraproject/IdentityAndConsentManagement/pull/287) |
| 0.3.0 | Deprecated | Yes, client-assertion lifetime cap | [#216](https://github.com/camaraproject/IdentityAndConsentManagement/pull/216) |
| 0.2.0 | Retired | N/A | N/A|

The ICM version lifecycle state for 0.x versions is assigned by ICM governance as a one-time exercise. Full SemVer does not apply to pre-1.0.0 versions.
```

Author: Not included, see previous comment. That is a separate discussion and not in the responsibility of the guideline.

BTW: 0.6.0 does not exist, Spring26 r4.2 is 0.5.0
I would also question to still support 0.3.0 (Spring25) - we would need to maintain it in case of security issues.

## 13. Exception mechanism

> Exceptions are time-bound compatibility waivers granted by governance:

Reviewer: ```suggestion
Exceptions are time-bound ICM-compatibility changes granted by governance:
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> 

Reviewer: ```suggestion

Exceptions shall be documented by Release Management using governance decision records with the following information:
 
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

- **Scope**: specific (API version, ICM version) pair, or a range.
- **Justification**: required — operational necessity, regulatory requirement, or security consideration.
- **Time bound**: explicit expiry date or condition.
- **Owner**: named API Sub Project or provider responsible for remediation by expiry.
- **Expiry**: automatic. No "ongoing exception" mechanism.

> Exceptions are the only mechanism by which a (API, ICM) pair can be deemed compatible despite violating the core rules in §4 and §6. They appear in the compatibility matrix with explicit annotation.

Reviewer: ```suggestion
Exceptions are the only mechanism by which a (API version, ICM version) pair can be considered ICM-compatible despite violating the ICM-compatibility rules. Exceptions will appear in the ICM-compatibility matrix with an explicit annotation.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

## 14. Open governance decisions

The following require WG agreement before this guideline is adopted:

1. **Exact durations** for the Supported and Deprecated state durations (§6.2).
2. **Transition to ICM 1.0.0** — what constitutes the scope baseline; when it is declared; who signs off.
3. **Signal vs. out-of-cycle policy** — which ICM changes can be out-of-cycle vs. must align with Signal.
4. **Exception grant process** — who requests, who approves, how documented.

> 5. **CAMARA-compliance statement content** — exact templates for API-specification compliance and ICM-implementation compliance; audit process.

Reviewer: ```suggestion
5. **ICM-compatibility statement**; define templates for API version ICM-compatibility and API implementation ICM-compatibility statements; audit process.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

6. **Interaction with GSMA certification** — does this decomposition map onto GSMA's processes; alignment discussion required.
7. **Handling of the ICM 0.x transition** — state assignments for existing 0.x versions; end date for 0.x support.

> 8. **Maximum number of concurrent non-Retired ICM majors** — whether to cap this to bound Provider operational complexity when ICM majors arrive in quick succession (for example, in a security-driven scenario), and how Retirement acceleration would be triggered if the cap is exceeded.

Reviewer: ```suggestion
8. **Maximum number of concurrent non-Retired major ICM versions** — whether to cap this to bound API Provider operational complexity when major ICM versions arrive in quick succession (for example, in a security-driven scenario), and how Retirement acceleration would be triggered if the cap is exceeded.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

9. **ICM ↔ Commonalities coupling** — ICM design info reaches API definitions through the CAMARA Commonalities API Design Guide (§3.4). Open question: how tightly must the Commonalities and ICM lifecycles couple? Options include mandating a new Commonalities release for each ICM design info change, moving ICM design artifacts into the ICM repository to decouple, or letting Commonalities itself declare a `x-camara-min-icm`. This is a coordination question between the ICM and Commonalities Working Groups.
10. **Example content for the §6.4 release-note tables** — Tables A and B in §6.4 are stubs awaiting concrete examples from ICM governance. The exact column layout will be refined once examples are filled in.

> 11. **Cross-major ICM-compatibility assessment — process and timing** — when in the meta-release cycle is the cross-major assessment performed for existing API versions (Signal? Sync? at the moment the new ICM major is announced?), who is responsible for the technical evaluation (API Sub Project? Release Management? ICM WG?), and how is the resulting decision recorded in the compatibility matrix.

Reviewer: ```suggestion
11. **Cross-major ICM-compatibility assessment — process and timing** — when in the meta-release cycle is the cross-major-ICM versions assessment performed for existing API versions (Signal? Sync? at the moment the new major ICM version is announced/released?), who is responsible for the technical evaluation (API Sub Project? Release Management? ICM WG?), and how is the resulting decision recorded in the ICM-compatibility matrix.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

---

## Appendix A: Key positions

- **ICM-compatibility has two distinct aspects** — design-time (API version, governed by CAMARA via Commonalities) and runtime (deployment, governed by API Provider). See §3.

> - **Meta-release is not the unit of compatibility.** Individual (API, ICM) pairs are, constrained by governance states and by the per-major binding rule in §4.

Reviewer: ```suggestion
- **A meta-release is not the unit of ICM-compatibility.** ICM-compatibility of individual (API version, ICM version) pairs are constrained by lifecycle states and by governance decisions/exceptions.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

- **`x-camara-min-icm`** is introduced as an OpenAPI extension carried by each API definition, independent of `x-camara-commonalities`.

> - **API Provider compliance is decomposed** into API-specification compliance and ICM-implementation compliance (§8).

Reviewer: ```suggestion
- **API Provider ICM-compatibility must cover both **API version ICM-compatibility** and **API deployment ICM-compatibility**.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

> - **The compatibility matrix is automated**, not hand-maintained.

Reviewer: ```suggestion
- **The ICM-compatibility matrix is derived by automation**, not hand-maintained.
```

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/e618c052324b1d60c6dcd62adcff08c532c04b8f

- **Signal/Sync cadence is a lead-time mechanism**, not a governance gate.
- **ICM design info reaches API definitions via Commonalities**, not directly (§3.4).

> ## Appendix B: What this guideline intentionally does not do

Reviewer: I would drop this section Appendix B to avoid confusion, at least in the final version and add a note accordingly.

Author: Applied in https://github.com/camaraproject/ReleaseManagement/pull/497/commits/5bca022

Both types of changes are captured in the ICM version according to standard SemVer rules:

- **Major version (1.x.y → 2.0.0)**: reserved for breaking changes that cannot be expressed additively. These may be due to ICM design info changes (for example, replacement of a mandatory security schema, an incompatible scope format, removal of a required claim), to ICM deployment info changes (for example, mandatory new authentication rules, an assertion lifetime cap that rejects existing Consumers), or both. Major ICM versions should be deliberately rare and driven by security or regulatory necessity.
- **Minor version change (1.x.y → 1.x+1.0)**: additive only. No breaking change, but minor change to ICM design info (preserves API version ICM-compatibility), or no breaking change, but minor change to ICM deployment info (preserves API deployment ICM-compatibility). For example, new optional flows, new optional claims, new recommendations permitted. Minor changes must not reject previously ICM-compatible API deployment behavior, as these may only apply in a major ICM version change.

> - **Patch version (1.2.3 → 1.2.4)**: documentation or defect corrections leading to patch API versions (maintain API version ICM-compatibility), and with no impact on API deployment ICM-compatibility. Example: update of the ICM mandatory text for the API version info.description field.

Author: "leading to" can be misread that every ICM triggers patch API versions. That's not the case, an ICM patch may be invisible to APIs.

```suggestion
- **Patch version (1.2.3 → 1.2.4)**: documentation or defect corrections that may require patch API versions to incorporate (e.g., update of the ICM mandatory text in the API version's `info.description` field), and with no impact on API deployment ICM-compatibility.
```

| ICM version | Breaking design info change | Affected construct | Impact on API definition |
|---|---|---|---|
| _2.0.0_ | _introduces new scope format X_ | _where it appears in an API definition_ | _what API teams need to do_ |

**Table B — Breaking ICM deployment info changes** (impacting API deployments; see §3.2). Used by API Providers and API Consumers to plan deployment updates.

| ICM version | Breaking deployment info change | Affected runtime behavior | Impact on API deployment |
|---|---|---|---|
| _0.3.0_ | _introduces 300s client-assertion lifetime cap_ | _which deployment behavior_ | _what Provider and Consumer need to do_ |

The exact column layout is subject to refinement once concrete examples are filled in (see §12).

## 6. API version ICM-compatibility - details

### 6.1 Minimum ICM version

An API version declares its minimum ICM version via the `x-camara-min-icm` field in its API definition:

```yaml
info:
  # The version of the CAMARA Identity and Consent Management (ICM)
  # specification that this API definition is built against.
  # This API version is compatible with higher ICM versions with the same major version number.
  x-camara-min-icm: 1.2.0
```

> This declared ICM version is the ICM version that the API definition relies on and which is Supported at the time of public release of this API version. 

Author: Here the notion of the "floor" got lost ... it reads like this is the (only) version the API definition is relying on. I propose to bring it back, here with "lowest" (other words possible as well):

```suggestion
This declared ICM version is the lowest ICM version that the API definition relies on and which is Supported at the time of public release of this API version.
```

- The value of `x-camara-min-icm` field must be a SemVer valid ICM version.
- It identifies the ICM version that this API is initially compatible with by design (e.g., `1.2.0`). The API version is guaranteed ICM-compatible with any higher ICM version that has the same major version number. ICM-compatibility with different higher major ICM versions must be assessed at ICM version public release and may be extended on governance decision as described below.
- Fixed at API public release. Never changes after the API version is released.

### 6.2 ICM governance responsibilities 

When a new major ICM version is publicly released, the ICM-compatibility of each API version in the ICM-compatibility matrix with respect to this new ICM version needs to be assessed by governance:
- API versions that do not rely on impacted ICM design info may be declared ICM-compatible with the new major ICM version through an explicit governance decision. This is recorded in the ICM-compatibility matrix.

> - API versions that do rely on the impacted ICM design info must release a new major API version with the new major ICM version in their `x-camara-min-icm` field. This new pair will be added to the ICM-compatibility matrix.

Author: I propose the "major" from the sentence. Not every ICM design info change forces a breaking change at the API level. Leaving the decision of the right SemVer version bump to the API Sub Project.

```suggestion
- API versions that do rely on the impacted ICM design info must release a new API version with the new major ICM version in their `x-camara-min-icm` field. This new pair will be added to the ICM-compatibility matrix.
```

Example: `x-camara-min-icm: 1.2.0` declared in an API definition means that the API version depends on ICM design info from ICM version 1.2.0 onward. The definition is guaranteed ICM-compatible with any Supported or Deprecated ICM version ≥ 1.2.0 with the same major version number (e.g. with 1.2.1, 1.3.0, but not with 2.0.0). Whether an API version remains ICM-compatible with a subsequent ICM version 2.0.0 is a governance assessment made at the time of the ICM 2.0.0 public version release.

### 6.3 API designer responsibilities

> To maintain ICM-compatibility of API versions, an API designer must release an update of the API version when it is no longer ICM-compatible. This incompatibility may be due to:

Author: Avoiding duplication of the same sentence as in 3.3:

```suggestion
Per §3.3, an API designer must release an update of the API version when it is no longer ICM-compatible. This incompatibility may be due to:
```

- Lifecycle state change of the referenced ICM version to Retired or Revoked. API designers must release a new API version referencing a Supported ICM version.
- Availability of a new major ICM version with changes that break either or both:
  - API version ICM-compatibility — affecting ICM design info (scope format, securitySchemes syntax, schemas, operations): API designers must release a new API version compatible with the new major ICM version.
  - API deployment ICM-compatibility — affecting ICM deployment info (auth flows, grant types, assertion format and lifetime, token processing): No need to release a new API version. Maintaining API deployment ICM-compatibility is the responsibility of API Providers and API Consumers.

### 6.4 Determining the `x-camara-min-icm` value

At each API version public release (new major, minor, or patch), the API Sub Project determines the value using the following formula:

```
x-camara-min-icm = max (
  lowest Supported ICM version at API version public release,
  lowest Supported ICM version containing all features this API's definition requires,

>   ICM version required by the Commonalities version declared in `x-camara-commonalities`

Author: I'm not sure why "lowest" was dropped here ... also Commalities does not require a specific version, but declares a version which is compatible with all guidelines in the releases version. E.g. in case of Spring26 that might be ICM 0.4.0 from Fall25, if there were no breaking changes and no changes from ICM brought into Commonalities (to be confirmed first ... might be also conditional if there were changes for specific APIs or features).

```suggestion
  lowest ICM version required by the Commonalities version declared in `x-camara-commonalities`
```

Therefore, this formula ensures that releasing a new version of such an API always raises its `x-camara-min-icm` to a Supported ICM version.

### 6.5 CAMARA validation support

A CAMARA linting rule, run at API version release, reads the current ICM lifecycle state (from the latest ICM release's notes) and verifies:
1. `x-camara-min-icm` is present in the API definition.
2. Its value is a syntactically valid SemVer 2.0 version.
3. The value refers to a Supported ICM version at API version public release.

This rule depends on the ICM lifecycle state being published in machine-readable form (see §5.4.1).

## 7. API deployment ICM-compatibility - details

This section describes the governance of API deployment ICM-compatibility (runtime).

An API definition might technically work against an ICM version that the compatibility matrix (§9) does not record as ICM-compatible — for example, a Retired or Revoked ICM version still running somewhere. Such pairings are outside CAMARA governance and are not supported by CAMARA.

### 7.1 API Provider responsibilities

To achieve ICM-compatibility of their API deployment, API Providers MUST:

1. Implement a Supported ICM version.

> 2. When a new major ICM version is published and is in Supported state, plan implementation of that ICM version. During the period in which the previous major ICM version remains in the Supported state, keep that major ICM version running in parallel, so that API Providers and Consumers can plan the migration of impacted API versions.

Author: just "migration planning" is for me too narrow as a reason to run two version in parallel, I would restore here the operational continuity as the main reason:

```suggestion
2. When a new major ICM version is published and is in Supported state, plan implementation of that ICM version. During the period in which the previous major ICM version remains in the Supported state, keep that major ICM version running in parallel, so that API versions targeting either Supported major ICM version remain deployable while API Providers and Consumers plan migration of impacted API versions.
```
